boilerplate/openshift/golang-osd-operator/standard.mk

```diff
@@ -153,11 +161,34 @@ go-generate:
 	${GOENV} go generate $(TESTTARGETS)
 	# Don't forget to commit generated files
 
+# go-get-tool will 'go install' any package $2 and install it to $1.
+define go-get-tool
+@{ \
+set -e ;\
+TMP_DIR=$$(mktemp -d) ;\
+cd $$TMP_DIR ;\
+go mod init tmp ;\
+echo "Downloading $(2)" ;\
+GOBIN=$(shell dirname $(1)) go install $(2) ;\
+echo "Installed in $(1)" ;\
+rm -rf $$TMP_DIR ;\
+}
+endef
+
+CONTROLLER_GEN = $(PWD)/bin/controller-gen
+CONTROLLER_GEN_VERSION = v0.8.0
+ifeq ($(USE_OLD_SDK), TRUE)
+CONTROLLER_GEN_VERSION = v0.3.0
+endif
+
+.PHONY: controller-gen
+controller-gen: ## Download controller-gen locally if necessary.
+	$(call go-get-tool,$(CONTROLLER_GEN),sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_GEN_VERSION))
+
 .PHONY: op-generate
-op-generate:
-	# The artist formerly known as `operator-sdk generate crds`:
+op-generate: controller-gen
```

Comment:
Sorry, I just clued into this.

Whereas this will make it easier to run locally, it conflicts with the way we're [version-controlling controller-gen from the bp backing image](https://github.com/openshift/boilerplate/blob/9030cbfacb46509b1e38eac089d3428b282fb166/config/build_image-v2.0.0.sh#L41-L45). Basically, someone using the golang-osd-operator convention will be ignoring that config and using the version you're setting up here, whereas someone using a different convention would get what they expect from that image.

What happens if you use controller-gen 0.8.0 on a pre-osdkv1 repo? If it works, and doesn't result in any *functional* differences*, then I would suggest updating there (so that everyone is using 0.8.0) and removing this whole section.

If it doesn't work -- if you truly need 0.3.0 for old osdk and 0.8.0 for new -- then I suggest updating the backing image to
- install _both_ binaries with version suffixes
- symlink 0.3.0 to the unadorned `controller-gen` name as the default for backward compatibility
- twiddle this makefile to assume 0.8.0 exists at the appropriate path and either re-symlink it or set a var that points to the right version-suffixed binary

Either way you're going to need a boilerplate backing image update, which I suggest you do in a separate PR as it involves some fiddly tag manipulation. See #204 for a recent example.

*it's okay if there are trivial deltas that can easily be committed to the consuming repo without affecting form or function

Reply:
Hi @2uasimojo we do need 0.3.0 for old, and 0.8.0 for new, cause it generate differences. 
PR updated, and I created https://github.com/openshift/boilerplate/pull/217